### ISSUE TITLE
Alignment crashing on macOS[BUG]

### ISSUE THREAD

**Issue by vigji** [SainsburyWellcomeCentre/amap-python#39]
I am trying to align the test dataset downloaded following the link of the aMAP wiki (I esported it as a sequence of tiff files - this seemed to me the way files should be organised in the data folder, but from my reading of the docs I could not find this explicitly). I created an environment `map_env`, installed via pip amap, and run the alignment from the prompt. 
Alignment starts, files seem to be loaded, RAM usage is high but manageable (peaks at 14 out of 16); after 5-10 mins from the last `Loading images` bar, it crushes saying only `Killed: 9` in the terminal.

I get the following associated log in the output directory:

```

**************  AMAP LOG  **************

Analysis carried out: 2020-04-21_11-34-35
Output directory: /Users/luigipetrucco/Desktop/out
Current directory: /Users/luigipetrucco
Version: 0.1.16

**************  GIT INFO  **************

Software does not appear to be in a git repository. Perhaps it was installed in some other way?

**************  COMMAND LINE ARGUMENTS  **************

Command: /Users/luigipetrucco/anaconda3/bin/amap 
Input arguments: ['/Users/luigipetrucco/Desktop/testBrain', '/Users/luigipetrucco/Desktop/out', '-x', '12.5', '-y', '12.5', '-z', '12.5', '--n-free-cpus', '2', '--debug']

**************  VARIABLES  **************

Namespace:
image_paths: /Users/luigipetrucco/Desktop/testBrain
registration_output_folder: /Users/luigipetrucco/Desktop/out
downsample_images: None
no_boundaries: False
registration_config: /Users/luigipetrucco/anaconda3/lib/python3.7/site-packages/amap/amap.conf.custom
sort_input_file: False
no_save_downsampled: False
affine_n_steps: 6
affine_use_n_steps: 5
freeform_n_steps: 6
freeform_use_n_steps: 4
bending_energy_weight: 0.95
grid_spacing: -10
smoothing_sigma_reference: -1.0
smoothing_sigma_floating: -1.0
histogram_n_bins_floating: 128
histogram_n_bins_reference: 128
x_pixel_um: 12.5
y_pixel_um: 12.5
z_pixel_um: 12.5
orientation: coronal
flip_x: False
flip_y: False
flip_z: False
rotation: x0y0z0
n_free_cpus: 2
debug: True
metadata: None
no_atlas: False
atlas: allen_2017_10um
install_path: /Users/luigipetrucco/.amap/atlas
download_path: None
no_amend_config: False


**************  LOGGING  **************

2020-04-21 11:34:35 AM - INFO - MainProcess fancylog.py:277 - Starting logging
2020-04-21 11:34:35 AM - INFO - MainProcess fancylog.py:278 - Not logging multiple processes
2020-04-21 11:34:35 AM - INFO - MainProcess cli.py:365 - Starting registration
2020-04-21 11:34:35 AM - DEBUG - MainProcess system.py:132 - Determining the maximum number of CPU cores to use
2020-04-21 11:34:35 AM - DEBUG - MainProcess system.py:141 - Number of CPU cores available is: 6
2020-04-21 11:34:35 AM - DEBUG - MainProcess system.py:166 - Setting number of processes to: 6
2020-04-21 11:34:35 AM - INFO - MainProcess main.py:86 - Preprocessing data for registration
2020-04-21 11:34:35 AM - INFO - MainProcess main.py:87 - Loading data
2020-04-21 11:34:35 AM - INFO - MainProcess batteryrunners.py:274 - pixdim[0] (qfac) should be 1 (default) or -1; setting qfac to 1
2020-04-21 11:34:35 AM - INFO - MainProcess brain_processor.py:80 - Loading raw image data
2020-04-21 11:34:35 AM - DEBUG - MainProcess brainio.py:72 - Data type is: directory of files
2020-04-21 11:34:35 AM - DEBUG - MainProcess utils.py:104 - Determining the maximum number of CPU cores to use
2020-04-21 11:34:35 AM - DEBUG - MainProcess utils.py:116 - Setting number of processes to: 6
```

**Comment by adamltyson** [comment]
Hi,

Thanks for giving this a go. I should note that [the original aMAP](https://github.com/SainsburyWellcomeCentre/aMAP) is now unsupported, and thus the demo data there is not designed to work easily with this python version.

Also, as with cellfinder. I have no specific plans to support macOS. That being said, there's no reason why it shouldn't work.

I think the problem is insufficient RAM, registering with the high resolution (10um) atlas uses a lot of RAM (somewhere around 25GB), so if you have 16GB, then it won't work. I'm planning on adding support for the 25um atlas (see https://github.com/SainsburyWellcomeCentre/amap-python/issues/2), which should use far less RAM, and run much quicker.

If you need this working on macOS, I can prioritise https://github.com/SainsburyWellcomeCentre/amap-python/issues/2.

**Comment by adamltyson** [comment]
Just following this up, having tested some stuff on a mac, and the registration fails (even with sufficient RAM). 

amap relies on [niftyreg](http://cmictig.cs.ucl.ac.uk/wiki/index.php/NiftyReg) to do the underlying registration, and on macOS, there seems to be an issue with the registration binaries. I didn't compile the macOS versions myself, so I'm not 100%, but I think they may need recompiling. 

As macOS is not supported, I won't be trying to fix this myself (I don't normally have access to a mac). If you want to try yourself, I'm happy to help.

Thanks

**Comment by vigji** [comment]
Ok, first thing then I will try on a Windows! 
Thanks a lot for these open tools and the quick rely, I'll try again and update the issue!

(On an unrelated note: any suggestion for how to deal with 1 hemisphere only data? Should I crop the reference or is this a scenario that is supported somehow?)

**Comment by vigji** [comment]
> Thanks for giving this a go. I should note that [the original aMAP](https://github.com/SainsburyWellcomeCentre/aMAP) is now unsupported, and thus the demo data there is not designed to work easily with this python version.

Yeah I realised that but I needed some test data and could not find any in amap_python!

**Comment by adamltyson** [comment]
I can share some data if you wanted. It's on my to-do list to add a function to download some demo data. 

For one hemisphere, this scenario is (currently) unsupported, but shouldn't be that difficult to add, it's also on my to-do list (https://github.com/SainsburyWellcomeCentre/amap-python/issues/1). Do you have cleared hemispheres, or something like that?

I'm fairly busy at the moment, but I can definitely help if you wanted to add some features for your data.

Adam

**Comment by adamltyson** [comment]
I'll respond to the hemispheres question on #1 

It should work on a 32GB windows machine (I've recently done it), but varies slightly on the size of the input data. I can't remember what the old demo data looks like. Can you just try cropping it closer to the surface of the brain, and see if that helps? 32GB is higher than it should be, I'll look into it (see https://github.com/SainsburyWellcomeCentre/amap-python/issues/42).

amap is designed to be run on full-resolution raw data (not the downsampled data from the old version). Prior to registration, the raw data is downsampled to the voxel resolution of the chosen atlas (10um by default), but the orientation of the raw data is maintained. For the actual registration, a copy of the atlas (and other files) is made, but they are transformed into the same orientation as your downsampled data (to best match prior to registration). These files are then not deleted, because they are used by packages such as [cellfinder](https://github.com/SainsburyWellcomeCentre/cellfinder) and [neuro](https://github.com/SainsburyWellcomeCentre/neuro). If your input data is exactly the same orientation as the provided atlas (in `~/.amap`), then they will just be copied, but that allows for portability of the results. Hope this makes sense!

**Comment by adamltyson** [comment]
Just following this up too, what was the answer to get the registration running on your machine?

**Comment by vigji** [comment]
Moving to a Windows machine with 110 GB of RAM :)
I will try to test again on the mac if ram usage can be a bit more efficient or with a downsampled atlas, but since it was failing without outputs I don't know if that was the issue.

**Comment by adamltyson** [comment]
As it is, I don't think it will work on macOS. I think the niftyreg binaries need to be recompiled. You're the first person (that I know of) to try this on a mac.

**Comment by vigji** [comment]
ok, I can give the recompiling a try!

**Comment by adamltyson** [comment]
Awesome, my knowledge of compiling C++ on macOS is precisely 0 (I don't write C++ or use a mac).

The issue (from memory) seems to be that they were compiled by Christian (who wrote the original Java version of amap), and the binaries are trying to call some library files on his machine. I assume it's something to the windows case, where we had to provide some windows redistributable binaries with the nityreg binaries.

Just to make things more difficult, the binaries are distributed as part of [imlib](https://github.com/adamltyson/imlib/tree/master/imlib/bin/nifty_reg). The original idea was to allow simple inverse transforms in other packages without depending on amap and all of its dependencies, but I think imlib may not be the best long term place for them.

**Comment by vigji** [comment]
To follow up on this - premise, I am absolutely not an expert on compiling C code either!

Imlib apparently has binaries for niftyreg, but they seems point to some funny user folder `/Users/Niedi/Documents/PhD/development/niftyreg/niftyreg_build150204/reg-lib/lib_reg_aladin.dyli`.
I managed to compile niftyreg from source and it seems to be working when I just run it from prompt - even using the exact same command for the affine transformation that is assembled in aMAP-python code and run through the `safe_execute_command`
For some funny reason that seems to have something to do with [this](https://stackoverflow.com/questions/35568122/why-isnt-dyld-library-path-being-propagated-here), I am having endless troubles to make the dynamic libraries accessible when I run code using `subprocess.check_call`.

I am posting here just to log the effort, but maybe if you are in contact with whoever compile these binaries they might have tips for how to solve such dynamic library linking problem

**Comment by vigji** [comment]
Ok, there's potentially a workaround - one can at runtime add the link to the dynamic libraries prepending `"export DYLD_LIBRARY_PATH=/path_to/niftyreg_install/lib && "` to the `cmd` string.
This is considerably ugly but apparently effective.
I can try to add it somehow to the `safe_execute_command` function, so would be a imlib pull request - unless you want to move the binaries from there - I agree that having the binaries distributed in imlib might be not that kosher

**Comment by adamltyson** [comment]
I'm afraid I'm no help, but I'll contact the person who originally compiled these binaries. I have a feeling the original amap never actually worked on macOS.

I'll outline the plan for all these repos, and see if you agree with where I think these binaries should go.

`amap` - Volume to volume registration using niftyreg. I'm thinking that other types of registration, e.g. 2D-3D should be taken care of by other packages (e.g. `slicereg`).
`cellfinder` - 3D cell detection in volumetric images. This will likely be extended to 2.5D (i.e. a series of 2D sections, not overlapping in Z).

`imlib` - This was originally shared image processing functions for `amap`, `cellfinder` and others. It has gradually expanded it's scope, and I'm keen to keep the dependencies simple.
`neuro` - As `imlib` was getting too big, this package exists for a lot of the `amap` and `cellfinder` post-processing (generating figures etc). The idea is that you can run the intensive `amap` and `cellfinder` processing on your workstation/cluster, but then work with the output directories (which will eventually be fully portable) on your laptop etc., without needing a great computer or a functional tensorflow installation. This should also get around a lot of the OS issues I'm having. `neuro` needs access to the binaries for applying existing transforms to new data etc.

There are some packages in the works, but these are the main ones. Basically I think `imlib` should be general functions, and `neuro` should be for brain-anatomy specific stuff.
 **TLDR: I think the binaries should be moved to neuro**

**Comment by adamltyson** [comment]
I think I misread your last comment, so it is now working, it's just a case of how to distribute this properly?

I'm happy with ugly if necessary. The path to the binaries could always be added to the config file upon installation if necessary.

**Comment by vigji** [comment]
> **TLDR: I think the binaries should be moved to neuro**

I think it makes sense as an organisation - and of course, it's your library! :D ; I see the point of having the binaries accessible from multiple places for the inverse transformation, just, as a matter of overall tidiness, it feels weird to have something like the `amap-python` package just an a tiny wrapper over binaries that are distributed separately. It seems that the heavy lifting of aMAP is done by the binaries, and it is overall quite light. Looking from my point of view - which of course is limited because of the partial use I have done of your code so far - it seems that it would be no such a big deal handling all  the transformations - including the later inverse transformations - through `amap-python`, which can be pretty safely be required also for tools that need to run from laptops, as there's no complicated/heavy dependencies inside it.

**Comment by vigji** [comment]
> I think I misread your last comment, so it is now working, it's just a case of how to distribute this properly?
> 
> I'm happy with ugly if necessary. The path to the binaries could always be added to the config file upon installation if necessary.

I was looking for a way of making it distributable; my current problem is that right now when compiling the binaries I think I have to link the location of the dynamic libraries files, and I am not sure how to keep that as a relative path from the binaries, might be trivial but I am out of my usual waters here :) I will try to see if I get to some reasonable solution and in case make a PR on imlib, but if you can get inputs from that person would be great - I am happy to recompile the binaries with their suggestions!

**Comment by adamltyson** [comment]
> Looking from my point of view - which of course is limited because of the partial use I have done of your code so far - it seems that it would be no such a big deal handling all the transformations - including the later inverse transformations - through amap-python

I agree, it's now at a point where `amap-python` is pretty light. The reason for not having the binaries in amap now, is to avoid a circular dependency. I'm not sure if this is an issue in the same way as circular imports, but amap will rely on neuro, so I'd like to avoid neuro relying on amap.

**Comment by vigji** [comment]
I see the point! Anyway, I think that the only way circular dependencies can be an issue is the way they could easily lead, in the general case, to circular imports - but something as self-contained as the niftyreg binary files should not be an issue as you don't import anything there and you just access them from `neuro` or `amap`

**Comment by vigji** [comment]
There's another issue: right now using the atlas seems to work mostly b/c it is always supposed to be in the same place; there's several functions in imlib that assume a specific location for the atlas, and some form of atlas register in the amap.conf file - I am already experiencing some troubles in the tests, i.e., it seems surprisingly complicated just use the downloaded 100um resolution atlas for the rest of the tests ( more on this on the PR).

What is the long term plan for the atlas register? the config file with the list of all atlases should live in only one place, so either `amap`, `imlib`, or the user root folder.

**Comment by adamltyson** [comment]
The functions do assume a specific location, but this is typically sourced [like this](https://github.com/adamltyson/imlib/blob/master/imlib/source/source_files.py#L20). However, this whole thing is messy, and is the product of lots of short-term fixes until the whole concept of multiple atlases and configs to support them is thought out (so probably now).

I'm not 100% sure where the config should be, originally I didn't want to force anything to be written to the home directory, I'm not sure if this relying on a default location could affect multiple-user installations.


----------
N.B a reminder to myself - when this stuff is changed, cellfinder will need to be changed to match. Particularly, cellfinder shouldn't be downloading it's own atlas.

**Comment by vigji** [comment]
> The functions do assume a specific location, but this is typically sourced [like this]

The problem with such sourcing it that if I am correct it will yield different results depending on the library you use it from!
The only 2 solutions I see for this problem are 
1. A default location for the a config file outside the folder, eg user home (but I do agree that it can be problematic from multiple-users installations)
2. Having `amap` (or `imlib` or...) as the single gateway for the entire suite to the available atlases on the machine; in this case you can have the conf there, and have all other libraries interacting with it to read such file. If this becomes the case, though, you might want to move to `amap` all that stuff from `imlib` that handles interfacing with such atlas (or from `amap` to `imlib`)

**Comment by adamltyson** [comment]
I think they should work (but I could be wrong). I use these functions to get the atlas from a few different libraries.

I think `2` is the best answer though. Currently there is code duplication between amap and cellfinder. I think `neuro` should be the gateway to all of these things.

**Comment by cniedwor** [comment]
Sorry for coming a bit late to the party but I didn't manage to look into it earlier.
Regarding amap on mac, I thought I had tested it on another machine, but since it is linked by absolute path, I must obviously be wrong here.

For compilation from source, there are detailed instruction in the wiki in the section "compiling from source"
https://github.com/SainsburyWellcomeCentre/aMAP/wiki/Compiling-from-source

potentially check the options by running ccmake instead of cmake and see whether you can change absolute/relative linking options here for distribution. I won't have time to properly look into this before the week-end at the earliest as I have changed careers, so don't have the build environment readily available right now.

More detailed options are also available at
http://cmictig.cs.ucl.ac.uk/wiki/index.php/NiftyReg_install

Hope this helps.
Best,
Christian

**Comment by adamltyson** [comment]
Thanks Christian. 

It would be great to get amap working on macOS, because then the whole pipeline will work across windows/mac/linux.

Not much I can do though as I don't have a mac. @vigji have you had any success?

**Comment by cniedwor** [comment]
So I have just reactivated my old mac from the lab running on 10.10.5, downloaded the original amap binaries from the current imlib master branch, deleted the referenced .dyli files you showed from my system (yes they did still exist at that path) and ran a transformation using reg_aladin from the commandline - which finished without any errors.

So my obvious question would be - how do I replicate this one? Does it only happen when you run it from the new python interface?

Best,
Christian

**Comment by vigji** [comment]
Hi @cniedwor,
I recompiled just yesterday the binaries following the links and recommendations you send and I managed to have them not to point to the .dyld libraries anymore. Unfortunately that discussion happened in the imlib PR and I did not updated the discussion here, sorry!

Problem seems to be solved. Thanks again for the links and for the double check :)

**Comment by adamltyson** [comment]
Reopening this, I don't think the new macOS binaries are fully portable. Running [amap on macOS (10.13) on Travis](https://travis-ci.com/github/SainsburyWellcomeCentre/amap-python/jobs/327954868) gives this error:

```
       except SafeExecuteCommandError as err:

            raise RegistrationError(

>               "Affine registration failed; {}".format(err)

            )

E           imlib.general.exceptions.RegistrationError: Affine registration failed; 

E           -------------------------

E           Process failed:

E            dyld: lazy symbol binding failed: Symbol not found: ____chkstk_darwin

E             Referenced from: /Users/travis/miniconda/envs/test-environment/lib/python3.7/site-packages/imlib/bin/nifty_reg/osX/reg_aladin (which was built for Mac OS X 10.15)

E             Expected in: /usr/lib/libSystem.B.dylib

E           

E           dyld: Symbol not found: ____chkstk_darwin

E             Referenced from: /Users/travis/miniconda/envs/test-environment/lib/python3.7/site-packages/imlib/bin/nifty_reg/osX/reg_aladin (which was built for Mac OS X 10.15)

E             Expected in: /usr/lib/libSystem.B.dylib
```

I don't know if it's a portablilty issue, or a macOS version issue.

**Comment by vigji** [comment]
Oh wow. I'm sorry, I have no idea!

**Comment by adamltyson** [comment]
Confusingly, this doesn't always fail, see: https://travis-ci.com/github/SainsburyWellcomeCentre/amap-python/jobs/328603344

**Comment by adamltyson** [comment]
I just tested registration on a mac (version 10.14.6), and it fails, trying to find `/usr/local/opt/libpng/lib/libpng16.16.dylib`

**Comment by vigji** [comment]
Uhm, interesting. It does ask for the location of libpng during the compilation. Do you have that file anywhere else on your machine? I had it in that location before using aMAP...

**Comment by adamltyson** [comment]
I don't know. It's my partners macbook, and it has nothing in terms of scientific/development software. I don't even know where to look on a mac.

**Comment by vigji** [comment]
I see. Well, I have no expertise for a systematic compiling that take into account different oc versions...I guess that people who run in such problems can just be pointed to the instructions for recompiling

**Comment by adamltyson** [comment]
Looks to be ok if users install `libpng`, see #53 

Thanks for testing all this stuff @vigji !

Closing this now. I still need to work out to what extent I will support macOS however.

**Comment by adamltyson** [comment]
Definately system dependent though. Sometimes I still get `dyld: lazy symbol binding failed: Symbol not found: ____chkstk_darwin`
